File: day22/sporifica_virus2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <stdlib.h>
#include "array2d.h"

using namespace std;

struct v {
    int x;
    int y;
};

void printg(array2d<char> &g) {
    for (int x = 0; x < g.x_dim(); x++) {
        for (int y = 0; y < g.y_dim(); y++) {
            printf("%c ", g(x,y));
        }
        printf("\n");
    }
    printf("\n");
}

const char turn_right(const char d) {
    switch (d) {
        case 'u': return 'r';
        case 'r': return 'd';
        case 'd': return 'l';
        default : return 'u';
    }
}

const char turn_rev(const char d) {
    switch (d) {
        case 'u': return 'd';
        case 'r': return 'l';
        case 'd': return 'u';
        default : return 'r';
    }
}

const struct v d2move(const char d) {
    struct v move = {0, 0};
    switch (d) {
        case 'u': move.x = -1; break;
        case 'r': move.y = 1;  break;
        case 'd': move.x = 1;  break;
        case 'l': move.y = -1; break;
    }
    return move;
}

const char turn_left(const char d) {
    switch (d) {
        case 'u': return 'l';
        case 'l': return 'd';
        case 'd': return 'r';
        default : return 'u';
    }
}

int main(int argc, char *argv[]) {
    const int SIZE = 501;
    int iter = atoi(argv[1]);
    array2d<char> g(SIZE, SIZE, '.');

    // parse the input
    ifstream infile(argv[2]);
    string line;
    int lineno = 0;
    if (infile.is_open()) {
        while (getline(infile, line)) {
            size_t len = line.size();
            size_t offset = (SIZE - len)/2;
            size_t row = lineno + offset;
            for (size_t i = 0; i < len; i++) {
                size_t col = i + offset;
                g(row, col) = line[i];
            }
            lineno++;
        }
    }

    int cnt = 0;
    char d = 'u';
    int row = SIZE/2;
    int col = SIZE/2;
    for (int it = 0; it < iter; it++) {
        const char val = g(row,col);
        if (val == '.') {
            d = turn_left(d);
            g(row,col) = 'W';
        } else if (val == '#' ) {
            d = turn_right(d);
            g(row,col) = 'F';
        } else if (val == 'W' ) {
            g(row,col) = '#';
            cnt++;
        } else {
            d = turn_rev(d);
            g(row,col) = '.';
        }
        struct v move = d2move(d);
        row += move.x;
        col += move.y;
    }

    printf("result2: %d\n", cnt);
}

File: day22/sporifica_virus2.py
#!/usr/bin/env python3
from sys import argv

def emptyg(size):
    return [ [ '.' for y in range(size) ] for x in range(size) ]

def printg(g):
    for row in range(len(g)):
        for col in range(len(g)):
            print('{} '.format(g[row][col]), end='')
        print('')
    print('')

def turn_right(d):
    if d == 'u':
        return 'r'
    elif d == 'r':
        return 'd'
    elif d == 'd':
        return 'l'
    else:
        return 'u'

def turn_left(d):
    if d == 'u':
        return 'l'
    elif d == 'l':
        return 'd'
    elif d == 'd':
        return 'r'
    else:
        return 'u'

def turn_rev(d):
    if d == 'u':
        return 'd'
    elif d == 'r':
        return 'l'
    elif d == 'd':
        return 'u'
    else:
        return 'r'

def d2move(d):
    if d == 'u':
        return -1,0
    elif d == 'r':
        return 0,1
    elif d == 'd':
        return 1,0
    else:
        return 0,-1

ITER = int(argv[1])
SIZE = 501
g = emptyg(SIZE)

# parse the input
filename = argv[2]
lineno = 0
with open(filename) as f:
    for line in f:
        line = line.rstrip()
        offset = int((SIZE - len(line)) / 2)
        row = lineno + offset
        for i in range(len(line)):
            col = i + offset
            g[row][col] = line[i]
        lineno += 1

# run ITER bursts
cnt = 0
d = 'u'
row,col = int(SIZE/2), int(SIZE/2)

for it in range(ITER):
    val = g[row][col]
    if val == '.':
        d = turn_left(d)
        g[row][col] = 'W'
    elif val == '#': 
        d = turn_right(d)
        g[row][col] = 'F'
    elif val == 'W':
        g[row][col] = '#'
        cnt += 1
    else:
        d = turn_rev(d)
        g[row][col] = '.'

    drow, dcol = d2move(d)
    row += drow
    col += dcol

print("result2:", cnt)
